### ISSUE TITLE
styled-components unnecessarily warns about multiple instances in some cases

### ISSUE THREAD

**Issue by joukosaastamoinen** [styled-components/styled-components#3932]
Currently, if there are multiple instances of styled-components on the same page, it logs this warning:

```
It looks like there are several instances of 'styled-components' initialized in this application. This may cause dynamic styles to not render properly, errors during the rehydration process, a missing theme prop, and makes your application bigger without good reason.

See https://s-c.sh/2BAXzed for more info.
```

In most cases, this is really helpful. 👍 But it's unnecessary if styled-components is configured correctly like this:

1. Use [`SC_ATTR`](https://styled-components.com/docs/advanced#avoiding-conflicts-with-thirdparty-styles-and-scripts) to ensure that `<style>` tags don't get mixed up between styled-components instances
2. Namespace CSS selectors (with eg. [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components) or with styled-components v6's namespace feature)

If styled-components _is_ configured correctly like that, then it seems wrong to display the warning.

I propose changing the warning display logic to this: only display the warning if multiple instances of styled-components are being used with the same `SC_ATTR` setting and with the same namespace.

Does that sound sensible?

I could look into implementing this if people agree that this would be a good change.

IMO having multiple instances of styled-components on the same page is a legit and important use case. It is not always a result of a misconfigured build or dependency resolution done wrong.

**Comment by probablyup** [comment]
The issue is that multiple versions of the same runtime are being loaded, which will definitely cause problems because there will be a race condition on which instance has all the styles. In a micro-frontend environment you need to serve SC once on a global in the window context and configure your micros to reference the browser global or use module federation or something like that to ensure SC is only loaded once and the reference is shared between the child React trees.

**Comment by nevcos** [comment]
@quantizor One of the [MFE best practices](https://micro-frontends.org/#:~:text=interface%20to%20others.-,Isolate%20Team%20Code,-Don%E2%80%99t%20share%20a) is to:
> **Isolate Team Code**
> Don’t share a runtime, even if all teams use the same framework. Build independent apps that are self contained. Don’t rely on shared state or global variables.

What you're suggesting breaks this fundamental rule, making MFEs not independent and harder to upgrade.

**Comment by quantizor** [comment]
If you can think of a way to make this library work such that multiple instances do not clash, I welcome it. One of the bigger issues is around SSR rehydration, the library scoops up everything that has been rendered via SSR and hydrates it into the runtime cache so we don't reinject styles. If instances don't have shared access to this cache, they are bound to overlap and cause issues.
